runtime_builders/publish_runtime.py

```diff
@@ -0,0 +1,166 @@
+]
+
+TAG_REGEX = '(?<=:|@)(.*)'
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--infile', '-i',
+                        help='templated cloudbuild config file.',
+                        required=True)
+    parser.add_argument('--bucket', '-b',
+                        help='GCS bucket to publish runtime to',
+                        default='runtime-builders')
+    parser.add_argument('--language',
+                        help='the language associated with this builder',
+                        required=True,
+                        choices=LANGUAGES)
+    args = parser.parse_args()
+
+    if args.language not in LANGUAGES:
+        logging.error('Invalid language \'{0}\' specified! '
+                      'Options:'.format(args.language))
+        logging.error(LANGUAGES)
+        return 1
+
+    templated_file = _resolve_tags(args.infile)
+    _publish_to_gcs(templated_file, args.language, args.bucket)
+
+
+def _resolve_tags(config_file):
+    """
+    Given a templated YAML cloudbuild config file, parse it, resolve image tags
+    on each build step's image to the corresponding digest, and write new
+    config with fully qualified images to temporary file for upload to GCS.
+
+    Keyword arguments:
+    config_file -- string representing path to
+    templated cloudbuild YAML config file
+
+    Return value:
+    path to temporary file containing fully qualified config file, to be
+    published to GCS.
+    """
+    with open(config_file, 'r') as infile:
+        try:
+            config = yaml.round_trip_load(infile)
+
+            for step in config.get('steps'):
+                image = step.get('name')
+                templated_step = _resolve_tag(image)
+                step['name'] = templated_step
+
+            s = yaml.round_trip_dump(config)
+
+            fd, ofile = tempfile.mkstemp()
+            with open(ofile, 'w') as outfile:
+                outfile.write(s)
+            os.close(fd)
+            return ofile
+        except yaml.YAMLError as e:
+            logging.error(e)
+            sys.exit(1)
+
+
+def _resolve_tag(image):
+    """
+    Given a path to a tagged Docker image in GCR, replace the tag with its
+    corresponding sha256 digest.
+    """
+    if ':' not in image:
+        logging.error('Image \'{0}\' must contain explicit tag or '
+                      'digest!'.format(image))
+        sys.exit(1)
+    elif 'sha256' in image:
+        return image
+    else:
+        m = regex.search(TAG_REGEX, image)
+        if m:
+            target_tag = m.string[m.start():m.end()]
+            base_image = m.string[0:m.start()-1]
+        else:
+            logging.error('Error when parsing image tag!')
+            logging.error(e)
+            sys.exit(1)
+
+    command = ['gcloud', 'beta', 'container', 'images',
+               'list-tags', base_image, '--format=json']
+
+    try:
+        output = subprocess.check_output(command)
+        entries = json.loads(output)
```

Comment:
I can, but that can be a ton of info since it will display every image digest (even the ones without tags). might make more sense to just log the entries for tagged images?